kep/115-Subgroup-support/README.md

```diff
@@ -0,0 +1,202 @@
+
+LeaderWorkerSet supports the leader not requesting TPU resources. This raises a problem when determining the values of `subgroup-index` and `subgroup-worker-index`. Suppose we have two TPU slices with two hosts each. Since the leader doesn’t request TPU resources, there will be four workers + the leader, meaning that one of the workers will have a worker index of four. Because of the way the new labels are calculated, this worker will have a subgroup-index of two, creating three subgroup indices (0,1,2) even though there are only two TPU slices.
+
+If the leader does not request TPU resources, then the labels will have the following values
+`leaderworkerset.gke.io/subgroup-index = (workerIndex - 1) / subGroupSize`
+`leaderworkerset.gke.io/subgroup-worker-index = (workerIndex - 1) % subGroupSize`
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+The goal is to ensure that we don't accept enhancements with inadequate testing.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+[ ] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+
+#### Unit Tests
+
+<!--
+In principle every added code should have complete unit test coverage, so providing
+the exact set of tests will not bring additional value.
+However, if complete unit test coverage is not possible, explain the reason of it
+together with explanation why this is acceptable.
+-->
+
+<!--
+Additionally, try to enumerate the core package you will be touching
+to implement this enhancement and provide the current unit coverage for those
+in the form of:
+- <package>: <date> - <current test coverage>
+
+This can inform certain test coverage improvements that we want to do before
+extending the production code to implement this enhancement.
+-->
+
+Will be added to ensure that the annotations are injected into the leader and worker pods, and to verify that the new values of TPU hostnames and TPU worker ID are injected. 
+
+#### Integration tests
+
+<!--
+Describe what tests will be added to ensure proper quality of the enhancement.
+
+After the implementation PR is merged, add the names of the tests here.
+-->
+
+- The pod-webhook integration tests should test
+  - Pod Affinity/Anti-Affinity is injected properly in leader and workers
+  - The new labels are only added if `SubGroupSize` is set
+  - The expected TPU environment variables are injected given different combinations of leader requests resources and subgroup sizes. 
+
```

Comment:
e2e tests can be added as well if we don't request TPU resources just to test the subgrouping and exclusive placement right?

Reply:
I added to test the subgrouping, but can e2e be added for exclusive placement? I don't see an existing e2e test for that

Reply:
e2e test means we will really schedule the pods into the nodes. As long as there are more than 1 nodes, then this should be possible right? even if only one nodes, we could still verify the pods have correct labels, and a certain pod is scheduled while others are pending?